### PR TITLE
add cyclic CD solver + fix GD for sparse data

### PR DESCRIPTION
**Pull request by agramfort**
@mathurinm if you have time to review ...

cc @josephsalmon

### CODE REVIEW

**Comment by agramfort** [comment]
teaser...

Screenshot 2021-04-29 at 14 40 14

File: solvers/cd.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+    def njit(f):  # noqa: F811
+        return f
+
+
+@njit
+def _newton_step_size(X, exp_yXw, j, lmbd):
+    hess_jj = 0.
+    for i in range(len(X)):
+        hess_jj += X[i, j]**2 * exp_yXw[i] / (1 + exp_yXw[i]**2)
```
Comment (mathurinm) [inline solvers/cd.py:21]: I believe the denominator should be `(1 + exp_yXw[i]) ** 2`

It's the second derivative of `ln(1 + exp((u))`, first derivative is sigmoid(u) and second is `sigmoid(u) * (1 - sigmoid(u))`

File: solvers/cd.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    return 1 / (hess_jj + lmbd)
+
+
+@njit
+def _newton_step_size_sparse(X_data, X_indices, X_indptr, exp_yXw, j, lmbd):
+    start, end = X_indptr[j:j+2]
+    hess_jj = 0.
+    for ind in range(start, end):
+        i = X_indices[ind]
+        hess_jj += X_data[ind]**2 * exp_yXw[i] / (1 + exp_yXw[i]**2)
```
Comment (mathurinm) [inline solvers/cd.py:31]: ditto

File: solvers/cd.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        else:
+            # use Fortran order to compute gradient on contiguous columns
+            self.X = np.asfortranarray(X)
+
+        # Make sure we cache the numba compilation.
+        self.run(1)
+
+    def _get_lipschitz_csts(self):
+        if sparse.issparse(self.X):
+            L = np.array((self.X.multiply(self.X)).sum(axis=0)).squeeze() / 4
```
Comment (mathurinm) [inline solvers/cd.py:59]: if scipy is a dependency we can use scipy.sparse.linalg.norm(X, axis=0)

**Comment by agramfort** [comment]
thanks !

>

File: solvers/python_gd.py [hunk 4]
```diff
@@ -13,19 +14,48 @@ def set_objective(self, X, y, lmbd):
     def run(self, n_iter):
         n_samples, n_features = self.X.shape
 
-        L = (np.linalg.norm(self.X) ** 2 / 4) + self.lmbd
+        L = self.compute_lipschitz_cste()
```
Comment (mathurinm) [inline solvers/python_gd.py:17]: cste sounds French
```suggestion
        L = self.compute_lipschitz_cst()
```

File: solvers/python_gd.py [hunk 5]
```diff
@@ -13,19 +14,48 @@ def set_objective(self, X, y, lmbd):
...
-            grad = -np.dot(X.T, (y * temp)) + lmbd * w
+            grad = -(X.T @ (y * temp)) + lmbd * w
             w -= step * grad
 
         self.w = w
 
     def get_result(self):
         return self.w
+
+    def compute_lipschitz_cste(self, max_iter=100):
```
Comment (mathurinm) [inline solvers/python_gd.py:34]:
```suggestion
    def compute_lipschitz_cst(self, max_iter=100):
```

File: solvers/python_gd.py [hunk 6]
```diff
@@ -13,19 +14,48 @@ def set_objective(self, X, y, lmbd):
...
+        if not sparse.issparse(self.X):
+            return np.linalg.norm(self.X, ord=2) ** 2 / 4 + self.lmbd
+
+        n, m = self.X.shape
```
Comment (mathurinm) [inline solvers/python_gd.py:38]: I discovered `scipy.sparse.linalg.svds`  (sparse truncated svd, uses Lanczos method which is faster than Power method)
`L = svds(X, k=1)[1][0] ** 2`

**Comment by agramfort** [comment]
thanks @mathurinm for the reviews !

**Pull request merged**